File: src/cpp/cdm/engine/SECondition.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/engine/SECondition.h"
#include "cdm/substance/SESubstanceManager.h"
#include "cdm/patient/conditions/SEPatientCondition.h"
#include "cdm/system/environment/conditions/SEEnvironmentCondition.h"
#include "cdm/io/protobuf/PBConditions.h"

SECondition::SECondition(Logger* logger) : Loggable(logger)
{
  m_Comment="";
}

SECondition::~SECondition()
{
  m_Comment = "";
}

void SECondition::Clear()
{
  m_Comment = "";
}

bool SECondition::SerializeToString(std::string& dst, eSerializationFormat fmt) const
{
  return PBCondition::SerializeToString(*this, dst, fmt);
}
SECondition* SECondition::SerializeFromString(const std::string src, eSerializationFormat fmt, const SESubstanceManager& subMgr)
{
  return PBCondition::SerializeFromString(src, fmt, subMgr);
}

std::string SECondition::GetComment() const
{
  return m_Comment;
}
void SECondition::SetComment(const std::string& comment)
{
  m_Comment = comment;
}
bool SECondition::HasComment() const
{
  return m_Comment.empty()?false:true;
}
void SECondition::InvalidateComment()
{
  m_Comment = "";
}

std::string SECondition::ToJSON() const
{
  std::string s;
  if (!SerializeToString(s, eSerializationFormat::VERBOSE_JSON))
    Error("Unable to serialize condition to json");
  return s;
}
std::string SECondition::ToString() const
{
  return PrettyPrint(this->ToJSON());
}

std::string SECondition::PrettyPrint(const std::string& str)
{
  return pulse::cdm::PrettyPrint(str, ePrettyPrintType::Condition);
}

File: src/csharp/pulse/cdm/engine/SECondition.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public abstract class SECondition
  {
    protected string comment;

    public SECondition()
    {
      comment = null;
    }

    public void Copy(SECondition other)
    {
      Clear();
      comment = other.comment;
    }

    public virtual void Clear()
    {
      comment = null;
    }

    public bool HasComment()
    {
      return !string.IsNullOrEmpty(comment);
    }
    public string GetComment()
    {
      return this.comment;
    }
    public void SetComment(string comment)
    {
      this.comment = comment;
    }
    public void InvalidateComment()
    {
      this.comment = null;
    }

    public abstract bool IsValid();

    public static string PrettyPrint(string msg)
    {
      return Logger.PrettyPrint(msg, Logger.ePrettyPrintType.Condition);
    }
  }
}
